xwin.c

```diff
@@ -2596,6 +2800,123 @@ handle_button_event(XEvent xevent, RD_BOOL down)
+									ctx->mod_in_progress |= mods[i].left_mod;
+								} else {
+									ctx->mod_in_progress = 0;
+									ctx->mod_allowed = ctx->mod_supplied;
+								}
+							} else if (ks == mods[i].right_ks) {
+								if (ctx->mod_allowed & mods[i].right_mod) {
+									ctx->mod_allowed ^= mods[i].right_mod;
+									ctx->mod_in_progress |= mods[i].right_mod;
+								} else {
+									ctx->mod_in_progress = 0;
+									ctx->mod_allowed = ctx->mod_supplied;
+								}
+							}
+						}
+					}
+				}
+
+			} else {
+				ctx->mod_in_progress = 0;
+				ctx->mod_allowed = ctx->mod_supplied;
+			}
+
+			if (ctx->mod_in_progress == ctx->mod_supplied) {
+				ctx->hit_ungrab_combo = 1;
+				logger(Keyboard, Debug, "%s: Seems that we have finally hit UNGRAB COMBO\n", __func__);
+			}
+
+			//if (i == 1 && text[0] == 'q') printf("Got 'q' press\n");
+
+			logger(Keyboard, Debug, "KeyPress after: mod_allowed = 0x%lx mod_supplied = 0x%lx mod_in_progress = 0x%lx hit_ungrab_combo = %ld\n", ctx->mod_allowed, ctx->mod_supplied, ctx->mod_in_progress, ctx->hit_ungrab_combo);
+
+			break;
+		case KeyRelease:
+			ks = XLookupKeysym((XKeyEvent *)ev, 0);
+
+			logger(Keyboard, Debug, "KeyRelease before: mod_allowed = 0x%lx mod_supplied = 0x%lx mod_in_progress = 0x%lx hit_ungrab_combo = %ld\n", ctx->mod_allowed, ctx->mod_supplied, ctx->mod_in_progress, ctx->hit_ungrab_combo);
+
+			if ((ks == XK_Alt_L) || (ks == XK_Alt_R) || (ks == XK_Shift_L) || (ks == XK_Shift_R) || (ks == XK_Control_L) || (ks == XK_Control_R) || (ks == XK_Super_L) || (ks == XK_Super_R)) {
+
+				if (ctx->mod_in_progress) {
+
+					for (i = 0; i < NUM_MODS; i++) {
+
+						if ((ks == mods[i].left_ks) || (ks == mods[i].right_ks)) {
+							/* Check for *_ANY first */
+							if (ctx->mod_in_progress & mods[i].any_mod) {
+								ctx->mod_allowed |= mods[i].any_mod;
+								ctx->mod_in_progress ^= mods[i].any_mod;
+							}  else {
+								if (ks == mods[i].left_ks) {
+									if (ctx->mod_in_progress & mods[i].left_mod) {
+										ctx->mod_allowed |= mods[i].left_mod;
+										ctx->mod_in_progress ^= mods[i].left_mod;
+									} else {
+										ctx->mod_in_progress = 0;
+										ctx->mod_allowed = ctx->mod_supplied;
+									}
+								} else if (ks == mods[i].right_ks) {
+									if (ctx->mod_in_progress & mods[i].right_mod) {
+										ctx->mod_allowed |= mods[i].right_mod;
+										ctx->mod_in_progress ^= mods[i].right_mod;
+									} else {
+										ctx->mod_in_progress = 0;
+										ctx->mod_allowed = ctx->mod_supplied;
+									}
+								}
+							}
+						}
+					}
+
+					if (ctx->hit_ungrab_combo && !ctx->mod_in_progress) {
+						return 1;
+					}
+				}
+			} else {
+				;
+			}
+
+			logger(Keyboard, Debug, "KeyRelease after: mod_allowed = 0x%lx mod_supplied = 0x%lx mod_in_progress = 0x%lx hit_ungrab_combo = %ld\n", ctx->mod_allowed, ctx->mod_supplied, ctx->mod_in_progress, ctx->hit_ungrab_combo);
+			break;
+		default:
+			break;
+	}
+
+	return 0;
+}
+
```

Comment:
Ideally, I would have preferred to see this split into two more specific functions that would more clearly communicate their purpose (bikeshedding: `ungrab_save_keypress`/`ungrab_modifiers_released`), but I'm not going to be picky about it.